src/Compilers/Core/Portable/Serialization/SerializationThreadPool.cs

```diff
@@ -0,0 +1,170 @@
+                                {
+                                    item.tcs.SetResult(((Func<object?>)item.action)());
+                                }
+                            }
+                            catch (OperationCanceledException ex)
+                            {
+                                item.tcs.TrySetCanceled(ex.CancellationToken);
+                            }
+                            catch (Exception ex)
+                            {
+                                item.tcs.TrySetException(ex);
+                            }
+                            finally
+                            {
+                                item.tcs.SetResult(true);
+                            }
+                        }
+                    });
+                    t.IsBackground = true;
+                    t.Start();
+                }
+
+                static void enqueue((Delegate, object?, TaskCompletionSource<object?>) item)
+                {
+                    // Enqueue the work. If there are currently fewer threads waiting
+                    // for work than there are work items in the queue, create another
+                    // thread. This is a heuristic, in that we might end up creating
+                    // more threads than are truly needed, but this whole type is being
+                    // used to replace a previous solution where every work item created
+                    // its own thread, so this is an improvement regardless of any
+                    // such inefficiencies.
+                    lock (s_queue)
+                    {
+                        s_queue.Enqueue(item);
+
+                        if (s_queue.Count <= s_availableThreads)
+                        {
+                            Monitor.Pulse(s_queue);
+                            return;
+                        }
+                    }
+
+                    // No thread was currently available.  Create one.
+                    createThread();
+                }
+
+                static bool tryDequeue(out (Delegate action, object? state, TaskCompletionSource<object?> tcs) item)
+                {
+                    // Dequeues the next item if one is available. Before checking,
+                    // the available thread count is increased, so that enqueuers can
+                    // see how many threads are currently waiting, with the count
+                    // decreased after. Each time it waits, it'll wait for at most
+                    // the idle timeout before giving up.
+                    lock (s_queue)
+                    {
+                        s_availableThreads++;
+                        try
+                        {
+                            while (s_queue.Count == 0)
+                            {
+                                if (!Monitor.Wait(s_queue, s_idleTimeout))
+                                {
+                                    if (s_queue.Count > 0)
```

Comment:
this is just for optimism right?  we'd only get into the if if the idle timeout expired, if htere had been an item added prior to that, we'd have been pulsed.  so this is just for the *very* narrow case of we timeout and just then an item is added?

Reply:
Right. That's my point. This is just for the very off chance that the timeout fired and it just do happens that right before that we had enqueued an item. But we were woken up for timeout, and not the pulse that occurred. 

It doesn't seem impossible. But it also seems very unlikely to ever happen.  So this feels like potential unnecessary complexity. I'm not saying to remove it. I'm just trying to gauge how important this is, and if my reading here is correct or not.

Reply:
I believe if we omit this check, it's possible for this thread to be notified of a new element, ignore it, and then not have another thread wake up to handle it instead.

Reply:
Hrmm... i'd like stephen to weigh in.  I would think teh Pulse woudl ensure that some waiting thread always wakes up and thus will see this value and process it.

Reply:
(where i'm not good at analyzing this sort of code is when timeouts are part of thigns.  i don't have a clear mental model of the lifetimes and how we ensure that we either kick off a thread if necessary, or we defer to an existing thread in a way that we're sure will ensure the work will always be picked up.

Reply:
Locks are not fair.  When thread 1 while holding the lock does a Pulse, it signals to the thread 2 at the front of the wait queue that it should become ready, but thread 2 still needs to reaquire the lock as part of successfully exiting from the Wait call, and thread 1 needs to exit the lock before that can happen.  When thread 2 does exit the lock, there's a race between thread 2 trying to acquire the lock and thread 3 coming along and trying to take the lock as well.  If thread 3 does acquire the lock before thread 2, it might dequeue an item, and there's then no guarantee that thread 2 would successfully dequeue an item when it sees Wait return true.  Hence the need for the second check; it's for correctness, not performance.

You can see this actually happening with a repro like:
```C#
using System;
using System.Threading;
using System.Threading.Tasks;

class Program
{
    static readonly ManualResetEventSlim mres1 = new();
    static readonly ManualResetEventSlim mres3 = new();
    static readonly object syncObj = new();

    static volatile int value;
    static volatile int count1;
    static volatile int count3;

    static void Main()
    {
        while (true)
        {
            value = 0;
            mres1.Reset();
            mres3.Reset();

            Task.WaitAll(
                Task.Run(() => // Thread #2
                {
                    lock (syncObj)
                    {
                        mres1.Set();
                        Monitor.Wait(syncObj);
                        if (value == 1) count1++;
                        else if (value == 3) count3++;
                        else throw new Exception(value.ToString());
                        Console.SetCursorPosition(0, 0);
                        Console.WriteLine("1: " + count1);
                        Console.WriteLine("3: " + count3);
                    }
                }),
                Task.Run(() => // Thread #1
                {
                    mres1.Wait();
                    lock (syncObj)
                    {
                        mres3.Set();
                        Monitor.Pulse(syncObj);
                        value = 1;
                    }
                }),
                Task.Run(() => // Thread #3
                {
                    mres3.Wait();
                    lock (syncObj)
                    {
                        value = 3;
                    }
                }));
        }
    }
}
```

Here Thread 2 takes the lock on syncObj, and only then is Thread 1 allowed to try to acquire it.  Thread 2 proceeds to Wait on the lock, which will enable Thread 1 to successfully acquire it.  Thread 1 then unblocks Thread 3, which now also tries to acquire the lock, and Thread 1 proceeds to pulse the syncObj in order to unblock Thread 2.  Before exiting the lock, Thread 1 sets an int value to be 1.  Now, if there weren't a race condition, Thread 2 would always win, and when it woke up from its Wait it would see that value is always 1.  If Thread 3 wins the race, it will set the value to 3, and so Thread 1 would see the int value as 3.  And if you run the above, you'll see that Thread 3 actually ends up winning most of the time.

Reply:
> Hence the need for the second check; it's for correctness, not performance.

That's great to know. Also, the above is exactly the sort of info that would be good to include in the pr. To me, it was not self evident and would be easy to screw up.

Can we add comments to this effect @sharwell ?

Reply:
Yes, although I need to adjust the description. The comment was written in this context (emphasis mine):

```
... no guarantee that thread 2 would successfully dequeue an item when it sees Wait return **true** ...
```

Since this check is located under the condition where `Wait` returns _false_, a similar condition can occur but the sequence is different.

Reply:
@CyrusNajmabadi details now added in a comment